### DISCUSSION TITLE
Libspdm implementation query

### DISCUSSION THREAD

**Discussion by PoojaGaur1708** [DMTF/libspdm#837] [Q&A] [answered]
Hello , i wanted to implement spdm over mctp over smbus . As I am using libspdm , i have writen my own application that will interact between libspdm and mctp . 

I called spdm_client_init().
In spdm_client_init(), called 
1. Libspdm_init_context
2. Libspdm_get_sizeof_required_scratch_buffer
3. Libspdm_register_device_io_func
4. Libspdm_register_transport_layer_func
5. Libspdm_register_device_buffer_func
6. Libspdm_set_data
7. Libspdm_init_connection

In libspdm_init_ connection ,
Returning error code 0xffffffff80010002
With libspdm_receive_spdm_response[O] (0x1240)

Can you pls tell me , why i am getting till error . Where I can get this error code description. 

And how can I run it successfully, if anything need to be added.

**Comment by jyao1** [comment]
It is unclear to me where the problem is, from your description.

The libspdm consumer reference is at https://github.com/DMTF/spdm-emu. The requester is at https://github.com/DMTF/spdm-emu/tree/main/spdm_emu/spdm_requester_emu. The responder is at https://github.com/DMTF/spdm-emu/tree/main/spdm_emu/spdm_responder_emu.

The status code definition is at https://github.com/DMTF/libspdm/blob/main/include/library/spdm_return_status.h

@PoojaGaur1708 
If you need more help, would you please post your code somewhere?
Or if you cannot post the code, please add some debug message to let us know the code flow.

I am sorry, but it is almost impossible to know what happened based upon such little information.

**Comment by PoojaGaur1708** [comment]
Hello Pls find the code below. Getting return Error in spdm_client_init Returning error code 0xffffffff80010002

**Reply by alokprasad** [reply]
I Dont see the code here?

**Comment by jyao1** [comment]
@PoojaGaur1708

At least one issue I found so far. The `libspdm_requester_receive_message()` is totally wrong.
```
message_size=malloc(sizeof(int));
*message_size=*(Rdata+1);

message=malloc(sizeof(void*));
*message=(__u8*)malloc(*message_size);
```

See below definition. You should **NOT** use **malloc** here, and you should fill the data to *message and *message_size directly.
```
 * @param  message                      A pointer to a destination buffer to store the message.
 *                                     The caller is responsible for having
 *                                     either implicit or explicit ownership of the buffer.
 *                                     On input, the message pointer shall be msg_buf_ptr from
 *                                     acquired receiver_buffer.
 *                                     On output, the message pointer shall be inside of
 *                                     [msg_buf_ptr, msg_buf_ptr + max_msg_size] from
 *                                     acquired receiver_buffer.

typedef libspdm_return_t (*libspdm_device_receive_message_func)(
    void *spdm_context, size_t *message_size, void **message,
    uint64_t timeout);
```

You may also see the implementation in spdm-emu at 
https://github.com/DMTF/spdm-emu/blob/main/spdm_emu/spdm_responder_emu/spdm_responder_spdm.c#L62

**Reply by PoojaGaur1708** [reply]
hello ,

I modified the function , remove the malloc  but still error is same.

Note : Rdata array contains all the data mctp+spdm . Can you pls tell me what should be in **message and m_send_receive_buffer**

as i received data in buffer (Rdata) as data[0] :0x00
data[1] :0x13 //Total No of bytes 
data[2] :0x0   //MCTP Packet header
data[3] :0xF
data[4] :0x0
data[5] :0x66
data[6] :0x1
data[7] :0x0
data[8] :0x0
data[9] :0xAD
data[10] :0x5
data[11] :0x7       //No. of SPDM bytes received  
data[12] :0x10     //spdm data
data[13] :0x4
data[14] :0x0
data[15] :0x0
data[16] :0x0
data[17] :0x1
data[18] :0x0




long unsigned int libspdm_requester_receive_message(void *spdm_context,
                                                 long unsigned int *message_size,
                                                void **message,
                                                long unsigned int timeout)
{
        __u8 len,Rdata[255];
        printf("===%s===:%d\n",__func__,__LINE__);
        len=sizeof(mctpPkt);
        sleep(2);               //Imp 
        for(int i=0;i<len;++i)
                {
                *(Rdata+i)=i2c_smbus_read_byte_data(file, i);
                m_send_receive_buffer[i]=*(Rdata+i);
                printf("data[%d] :0x%X\n",i,*(Rdata+i));

                }

        assert (*message == m_send_receive_buffer);
        m_send_receive_buffer_size = sizeof(m_send_receive_buffer);


        *message = m_send_receive_buffer;
        *message_size = m_send_receive_buffer_size;
        for(int i=0;i<*message_size;++i)
        {
               printf("Data Received: %x\n",*(__u8*)(*(message)+i));
        }
        printf("===%s===:%d\n",__func__,__LINE__);
        return LIBSPDM_STATUS_SUCCESS;
}

Error Get:
libspdm_receive_spdm_response[0] (0x1240): 
libspdm_receive_spdm_response[0] status - 0xffffffff80010002
===spdm_device_release_receiver_buffer===:52
spdm_client_init: 243
libspdm_init_connection - 0x80010002

**Comment by jyao1** [comment]
Where is your dump for below
```
    for(int i=0;i<*message_size;++i)
    {
           printf("Data Received: %x\n",*(__u8*)(*(message)+i));
    }
    printf("===%s===:%d\n",__func__,__LINE__);
```

Can you show the full log?

**Reply by PoojaGaur1708** [reply]
Pls find the full log.

===spdm_device_release_sender_buffer===:31
===spdm_device_acquire_receiver_buffer===:41
===spdm_device_acquire_receiver_buffer===:47 max_msg_size:482275864 buffer Address :0x782040
===libspdm_requester_receive_message===:110
===libspdm_requester_receive_message===:113 max_msg_size:0 buffer Address :0x782040
data[0] :0xFF
data[1] :0x13
data[2] :0x0
data[3] :0xF
data[4] :0x0
data[5] :0x66
data[6] :0x1
data[7] :0x0
data[8] :0x0
data[9] :0xAD
data[10] :0x5
data[11] :0x7
data[12] :0x10
data[13] :0x4
data[14] :0x0
data[15] :0x0
data[16] :0x0
data[17] :0x1
data[18] :0x0
data[19] :0x55
data[20] :0x55
data[21] :0x55
data[22] :0x55
data[23] :0x55
data[24] :0x55
data[25] :0x55
data[26] :0x55
data[27] :0x55
data[28] :0x55
data[29] :0x55
data[30] :0x55
data[31] :0x55
data[32] :0x55
data[33] :0x55
data[34] :0x55
data[35] :0x55
data[36] :0x55
data[37] :0x55
data[38] :0x55
data[39] :0x55
data[40] :0x55
data[41] :0x55
data[42] :0x55
data[43] :0x55
data[44] :0x55
data[45] :0x55
data[46] :0x55
data[47] :0x55
data[48] :0x0
data[49] :0x55
data[50] :0x55
data[51] :0x55
data[52] :0x55
data[53] :0x55
data[54] :0x55
data[55] :0x55
data[56] :0x55
data[57] :0x55
data[58] :0x55
Data Received: ff
Data Received: 13
Data Received: 0
Data Received: f
Data Received: 0
Data Received: 66
Data Received: 1
Data Received: 0
Data Received: 0
Data Received: ad
Data Received: 5
Data Received: 7
Data Received: 10
Data Received: 4
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 1
Data Received: 0
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 0
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 55
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
Data Received: 0
===libspdm_requester_receive_message===:133 max_msg_size:256 buffer Address :0x782040
libspdm_receive_spdm_response[0] (0x1240): 
libspdm_receive_spdm_response[0] status - 0xffffffff80010002
===spdm_device_release_receiver_buffer===:54
spdm_client_init: 244
libspdm_init_connection - 0x80010002

**Comment by jyao1** [comment, marked as answer]
I feel you should use `*message = &m_send_receive_buffer[10];`, since your MCTP message starts at byte 10.

However, why there is 0x7 in the middle?

```
data[10] :0x5
data[11] :0x7 //No. of SPDM bytes received
data[12] :0x10 //spdm data
```

@PoojaGaur1708 , you are **NOT** following https://www.dmtf.org/sites/default/files/standards/documents/DSP0275_1.0.0.pdf. 

Please read the spec. the length field should be absent.

**Reply by PoojaGaur1708** [reply]
thanks it resolve my this issue .

**Comment by PoojaGaur1708** [comment]
Hello ,

Now i am getting this error status return libspdm_init_connection - 0x80010007 which is LIBSPDM_STATUS_NEGOTIATION_FAIL.
Can you pls tell me how to resolve this . 

As i check the code , 
In libspdm_try_get_version() , we are passing session_ID as NULL in libspdm_receive_spdm_response(), and is_app_message as false in libspdm_receive_response(). which makes in function libspdm_receive_response() to return error code as LIBSPDM_STATUS_NEGOTIATION_FAIL.

      libspdm_zero_mem(spdm_response, spdm_response_size);
      status = libspdm_receive_spdm_response(spdm_context, **NULL**, &spdm_response_size,
                                           (void ****)&spdm_response);

 libspdm_return_t libspdm_receive_spdm_response(libspdm_context_t *spdm_context,
                                               const uint32_t *session_id,
                                               size_t *response_size,
                                               void **response)
{
    libspdm_session_info_t *session_info;
    libspdm_session_state_t session_state;

    if ((session_id != NULL) &&
        libspdm_is_capabilities_flag_supported(
            spdm_context, true,
            SPDM_GET_CAPABILITIES_REQUEST_FLAGS_HANDSHAKE_IN_THE_CLEAR_CAP,
            SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_HANDSHAKE_IN_THE_CLEAR_CAP)) {
        session_info = libspdm_get_session_info_via_session_id(
            spdm_context, *session_id);
        LIBSPDM_ASSERT(session_info != NULL);
        if (session_info == NULL) {
            return LIBSPDM_STATUS_INVALID_STATE_LOCAL;
        }
        session_state = libspdm_secured_message_get_session_state(
            session_info->secured_message_context);
        if ((session_state == LIBSPDM_SESSION_STATE_HANDSHAKING) &&
            !session_info->use_psk) {
            session_id = NULL;
        }
    }

    return libspdm_receive_response(spdm_context, session_id, false,
                                    response_size, response);
}

Output :
libspdm_receive_spdm_response[0] (0xff): 
0000: 10 04 00 00 00 01 00 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 
0020: 55 55 55 55 55 00 55 55 55 55 55 55 55 55 55 55 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
0040: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
0060: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
0080: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 20 84 e8 01 00 00 00 00 00 01 
===spdm_device_release_receiver_buffer===:52
libspdm_init_connection - 0x80010007
